doc/design/request_response_communication.md

```diff
@@ -0,0 +1,118 @@
+
+#### Typed API
+
+![typed API](diagrams/request_response/typed_api.svg)
+
+Since the `Response` is tied to a specific `Request`, the `loan` method takes a `Request` to populate the `Response` with the correct settings.
+Alternatively, a `Request` could have a `createResponse` method which returns a `Response` with the correct settings.
+
+#### Untyped API
+
+![untyped API](diagrams/request_response/untyped_api.svg)
+
+Similar to the the typed API, `loan` takes a pointer to a `RequestHeader` to populate the `ResponseHeader` with the correct settings.
+
+#### Client Port
+
+![client port](diagrams/request_response/client_port.svg)
+
+The `ClientPortData` is located in the shared memory and contain only the data but no methods to access them.
+`ClientPortUser` is the class providing the methods for the user access and `ClientPortRouDi` provides the
+interface RouDi needs to connect the client to the server and to cleanup the port resources.
+
+#### Server Port
+
+![server port](diagrams/request_response/server_port.svg)
+
+Similar to the Client Port, the Server Port has `ServerPortData` which is located in the shared memory and contain only the data but no methods to access them.
+`ServerPortUser` is the class providing the methods for the user access and `ServerPortRouDi` provides the
+interface RouDi needs to connect the client to the server once the server offers its service and to cleanup the port resources.
+
+It must be ensured that only one server with a given `ServiceDescription` can run at a time.
+
+#### Request/Response Header
+
+![rpc header](diagrams/request_response/request_response_header.svg)
+
+Since request and response need to encode different meta-information, we also need different header for the messages.
+The common data is aggregated in `RpcBaseHeader` which contains a `UniquePortId` to the port owning the `ClientChunkQueueData_t` and a sequence ID.
+The `UniquePortId` is used to identify the queue which receives the response and used as identifier for `ChunkDistributor::deliverToPort`.
+This method will iterate over all stored queues and matches the `ChunkQueueData::m_uniqueIdOfOwner`.
+As optimization, the `ChunkDistributor::deliverToPort` returns the index of the queue to do a fast lookup and downgrades to a full lookup if the port IDs don't match.
+```cpp
+class ChunkDistributor {
+    ...
+    cxx::expected<uint64_t, Error> deliverToPort(UniquePortId uniquePortId, uint64_t lastKnownQueueIndex);
+    ...
+};
+```
+
+The sequence ID is used to match a response to a specific request if multiple requests are invoke asynchronously.
+Depending on the client and server options, a request might be dropped or a server could have a worker pool
```

Comment:
How do we communicate a rejected request? What are the reasons for rejecting requests? E.g.

1.) busy - some queue full or otherwise resource starved
2.) access rights?
3.) wrong arguments?
Especially in a resource starved case, we may not be able to communicate failure to serve the request though (should be uncommon, however).

Reply:
Basically it is just a full queue with a `DISCARD_OLDEST_DATA` policy. Everything else should be handled by the server which means it could also response with a `REJECTED` message.